### DISCUSSION TITLE
Support dscp-based traffic bypass (such as bittorrent)

### DISCUSSION THREAD

**Discussion by mzz2017** [daeuniverse/dae#295] [General]
## Background

In cross-machine scenarios, pname matching will become ineffective. At this time, it will become difficult to match traffic such as BitTorrent.

We introduce dscp-based routing matching rules to solve this problem.

![](https://github.com/daeuniverse/dae/assets/30586220/9b7e873f-70d1-4a93-94ba-cc7786b88650)

Image is from <https://www.researchgate.net/figure/DSCP-field-in-the-header-of-IPv6-and-IPv4-packets_fig1_272538162>.

## Usage & Routing Format

```
routing {
  // decimal
  dscp(4) -> direct
  // hex
  dscp(0x4) -> direct
}
```

## Set dscp for the specific process on Windows

This section assumes you want to set a dscp to `qbittorrent.exe`.

First, launch the terminal (powershell):

1. Press `win+x`.
2. Select `Terminal (Admin)` or `PowerShell (Admin)`.

Second, set up a new QoS policy for qbittorrent.exe:

```powershell
New-NetQosPolicy -Name "qb" -AppPathNameMatchCondition "qbittorrent.exe" -PolicyStore ActiveStore -DSCPAction 4
```

In this example, dscp value 4 will be set.

Then, retrive the result to verify the success:

```powershell
Get-NetQosPolicy -PolicyStore "ActiveStore"
```

> If you want to delete it, use:
>
> ```powershell
> Remove-NetQosPolicy -Name "qb" -PolicyStore "ActiveStore"
> ```

## Set dscp in qBittorrent application

Setting DSCP is supported natively, but it has some issues and will miss some traffic, especially TCP traffic. This is because libtorrent set dscp after the **application protocol handshake** while proxy software split traffic at the very beginning the **TCP handshake**. At that moment, dscp is not set by qBittorrent yet.

You can find `peer-tos` or `peer-dscp` in the qBittorrent application advanced setting. You may need to pay attention that dscp is the value of tos divided by 4.

**Comment by sdandroid** [comment]
`New-NetQosPolicy -Name "qb"  -AppPathNameMatchCondition "qbittorrent.exe" -PolicyStore GPO:localhost  -DSCPAction 1 `
You can save it permanently or it will be gone if you reboot.
